### PR TITLE
Better management of the Page Visibility API

### PR DESCRIPTION
**Pull request by aberthet-gpsw**
This PR :

- updates the Page Visibility API in System and Device
- introduces two new properties for medias (videos and sounds of all origins): autoPause and autoResume

Medias are now paused/resumed when the window is getting out of focus/back in focus. Can be tested in any sample containing a media.

### CODE REVIEW

**Comment by sw-team-release-gpsw** [comment]
[CLA assistant check](https://cla.gopro.com/sign) <br/>All committers have signed the CLA.

File: src/audio/Playlist.js [hunk 1]
```diff
@@ -184,6 +200,9 @@ FORGE.Playlist.prototype._parseConfig = function(config)
 
     this._maxVolume = (typeof config.volume !== "undefined" && typeof config.volume.max === "number") ? FORGE.Math.clamp(config.volume.max, 0, 1) : 1;
     this._volume = (typeof config.volume !== "undefined" && typeof config.volume.default === "number") ? FORGE.Math.clamp(config.volume.default, 0, this._maxVolume) : FORGE.Math.clamp(1, 0, this._maxVolume);
+
+    this._autoPause = (typeof config.autoPause !== "undefined") ? !!config.autoPause : null;
+    this._autoResume = (typeof config.autoResume !== "undefined") ? !!config.autoResume : null;
```
Comment (rroux-gpsw) [inline src/audio/Playlist.js]: this._autoPause = (typeof config.autoPause == "boolean") ? config.autoPause : defaultValue;

File: src/audio/Sound.js [hunk 2]
```diff
@@ -1147,6 +1175,41 @@ FORGE.Sound.prototype.unmute = function()
...
+FORGE.Sound.prototype._onVisibilityChange = function()
+{
+    var status = document[FORGE.Device.visibilityState];
+
+    // What to do: Sound > Playlist > PlaylistManager > SoundManager
+    var autoPause = (typeof this._autoPause === "boolean") ? this._autoPause :
+                    (this._viewer.playlists.playlist !== null && typeof this._viewer.playlists.playlist.autoPause === "boolean") ? this._viewer.playlists.playlist.autoPause :
+                    (typeof this._viewer.playlists.autoPause === "boolean") ? this._viewer.playlists.autoPause :
+                    this._viewer.audio.autoPause;
+    var autoResume = (typeof this._autoResume === "boolean") ? this._autoResume :
```
Comment (rroux-gpsw) [inline src/audio/Sound.js]: Find better syntax. What happens on hotspots sound ?

File: src/audio/PlaylistManager.js [hunk 3]
```diff
@@ -273,6 +289,9 @@ FORGE.PlaylistManager.prototype._applyConfig = function(config)
 
         this._maxVolume = (typeof config.volume !== "undefined" && typeof config.volume.max === "number") ? FORGE.Math.clamp(config.volume.max, 0, 1) : 1;
         this._volume = (typeof config.volume !== "undefined" && typeof config.volume.default === "number") ? FORGE.Math.clamp(config.volume.default, 0, this._maxVolume) : FORGE.Math.clamp(1, 0, this._maxVolume);
+
+        this._autoPause = (typeof config.autoPause !== "undefined") ? !!config.autoPause : null;
+        this._autoResume = (typeof config.autoResume !== "undefined") ? !!config.autoResume : null;
```
Comment (rroux-gpsw) [inline src/audio/PlaylistManager.js]: Same as before :)
this._autoPause = (typeof config.autoPause == "boolean") ? config.autoPause : defaultValue;

File: src/audio/PlaylistTrack.js [hunk 4]
```diff
@@ -101,6 +117,9 @@ FORGE.PlaylistTrack.prototype._parseConfig = function(config)
     this._author = new FORGE.LocaleString(this._viewer, config.author);
     this._url = config.url;
     this._loop = config.loop || false;
+
+    this._autoPause = (typeof config.autoPause !== "undefined") ? !!config.autoPause : null;
+    this._autoResume = (typeof config.autoResume !== "undefined") ? !!config.autoResume : null;
```
Comment (rroux-gpsw) [inline src/audio/PlaylistTrack.js]: Same as before :)
this._autoPause = (typeof config.autoPause == "boolean") ? config.autoPause : defaultValue;

File: src/audio/PlaylistTrack.js [hunk 5]
```diff
@@ -449,7 +511,7 @@ Object.defineProperty(FORGE.PlaylistTrack.prototype, "loop",
     /** @this {FORGE.PlaylistTrack} */
     set: function(value)
     {
-        this._loop = Boolean(value);
+        this._loop = !!value;
```
Comment (rroux-gpsw) [inline src/audio/PlaylistTrack.js]: Why change this? the Boolean(var) syntax is the one used on the project from the start.

File: src/audio/Sound.js [hunk 6]
```diff
@@ -900,6 +919,9 @@ FORGE.Sound.prototype.play = function(position, loop, forceRestart)
         this._currentTime = /** @type {number} */ (position);
         this._playCount++;
 
+        this._viewer.onPause.add(this._onVisibilityChange, this);
+        this._viewer.onResume.add(this._onVisibilityChange, this);
```
Comment (rroux-gpsw) [inline src/audio/Sound.js]: So you add these listeners every time you call play method?

File: src/audio/Sound.js [hunk 7]
```diff
@@ -858,6 +874,9 @@ FORGE.Sound.prototype.play = function(position, loop, forceRestart)
             this._currentTime = /** @type {number} */ (position);
             this._playCount++;
 
+            this._viewer.onPause.add(this._onVisibilityChange, this);
+            this._viewer.onResume.add(this._onVisibilityChange, this);
```
Comment (rroux-gpsw) [inline src/audio/Sound.js]: So you add these listeners every time you call play method?

File: src/audio/Sound.js [hunk 8]
```diff
@@ -926,6 +948,12 @@ FORGE.Sound.prototype.stop = function(internal)
         this._playing = false;
         this._paused = false;
 
+        if (this._loop !== true)
+        {
+            this._viewer.onPause.remove(this._onVisibilityChange, this);
+            this._viewer.onResume.remove(this._onVisibilityChange, this);
```
Comment (rroux-gpsw) [inline src/audio/Sound.js]: I don't understand why the listener is binded to the pause and resume events of the viewer?

File: src/audio/SoundManager.js [hunk 9]
```diff
@@ -342,7 +358,7 @@ FORGE.SoundManager.prototype._applyConfig = function(config)
 {
     if(config !== null)
     {
-        this._enabled = typeof config.enabled !== "undefined" ? Boolean(config.enabled) : true;
+        this._enabled = (typeof config.enabled !== "undefined") ? !!config.enabled : true;
```
Comment (rroux-gpsw) [inline src/audio/SoundManager.js]: Why changing this syntax :( This is the closure compiler job to write code like this ;)

**Comment by aberthet-gpsw** [comment]
Ok, we're going back to a simpler model here, only with a global configuration in the sound manager, and removing the ones in the `playlists`, `lists`, `track` and `sound`. The configuration should be then shared between each track, ambient sound and hotspot sounds.

**Comment by aberthet-gpsw** [comment]
@bbaudel-gpsw @rroux-gpsw it should be good now, waiting for a review !

**Pull request merged**